# ElasticNet Regression: Complete Guide and Derivation

## Introduction to ElasticNet Regression

**ElasticNet Regression** is a regularized regression technique that combines both **L1 (Lasso)** and **L2 (Ridge)** regularization penalties. It was introduced to overcome the limitations of both Lasso and Ridge regression when used individually.

### Why ElasticNet?

1. **Ridge Regression (L2)**: Good at handling multicollinearity but keeps all features (doesn't perform feature selection)
2. **Lasso Regression (L1)**: Performs feature selection but can be unstable when features are highly correlated
3. **ElasticNet**: Combines the best of both worlds - performs feature selection like Lasso while maintaining stability like Ridge

### Key Advantages:

- **Feature Selection**: Like Lasso, ElasticNet can shrink coefficients to exactly zero
- **Grouped Selection**: When features are correlated, ElasticNet tends to select groups of correlated features
- **Stability**: More stable than Lasso in the presence of highly correlated features
- **Flexibility**: The `l1_ratio` parameter allows you to control the balance between L1 and L2 penalties

In [7]:
from sklearn.datasets import load_diabetes
from sklearn.linear_model import LinearRegression,Ridge,Lasso,ElasticNet
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score
import matplotlib.pyplot as plt
import numpy as np


## Load the Dataset

We'll use the **Diabetes dataset** from scikit-learn, which is a standard regression dataset with:
- **442 samples** 
- **10 features** (age, sex, BMI, blood pressure, and 6 blood serum measurements)
- **Target**: A quantitative measure of disease progression one year after baseline

In [10]:
X,y = load_diabetes(return_X_y=True)

In [12]:
X.shape, y.shape

((442, 10), (442,))

## Train-Test Split

We split the data into training (80%) and testing (20%) sets to evaluate model performance on unseen data.

In [17]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=2)

## Model Comparisons

We'll compare four different regression techniques to understand how ElasticNet improves over basic models.

### 1. Linear Regression (Baseline)

**Cost Function:**
$$J(\theta) = \frac{1}{2m} \sum_{i=1}^{m} (h_\theta(x^{(i)}) - y^{(i)})^2$$

Where:
- $m$ = number of training examples
- $h_\theta(x) = \theta^T x$ = hypothesis function
- No regularization penalty

**Characteristics:**
- Simple and interpretable
- No regularization
- Can overfit on high-dimensional data
- Sensitive to multicollinearity

In [18]:
# Linear Regression
linear_regressor_model = LinearRegression()
linear_regressor_model.fit(X,y)
y_pred = linear_regressor_model.predict(X)
r2_score(y,y_pred)

0.5177484222203499

### 2. Ridge Regression (L2 Regularization)

**Cost Function:**
$$J(\theta) = \frac{1}{2m} \sum_{i=1}^{m} (h_\theta(x^{(i)}) - y^{(i)})^2 + \alpha \sum_{j=1}^{n} \theta_j^2$$

Where:
- $\alpha$ = regularization parameter (controls strength of penalty)
- L2 penalty = $\alpha \sum_{j=1}^{n} \theta_j^2$ (sum of squared coefficients)

**Characteristics:**
- Adds L2 penalty to prevent overfitting
- Shrinks coefficients but never makes them exactly zero
- Good for handling multicollinearity
- All features remain in the model

In [19]:
# Ridge 
reg = Ridge(alpha=0.1)
reg.fit(X_train,y_train)
y_pred = reg.predict(X_test)
r2_score(y_test,y_pred)

0.45199494197195456

### 3. Lasso Regression (L1 Regularization)

**Cost Function:**
$$J(\theta) = \frac{1}{2m} \sum_{i=1}^{m} (h_\theta(x^{(i)}) - y^{(i)})^2 + \alpha \sum_{j=1}^{n} |\theta_j|$$

Where:
- $\alpha$ = regularization parameter
- L1 penalty = $\alpha \sum_{j=1}^{n} |\theta_j|$ (sum of absolute values of coefficients)

**Characteristics:**
- Adds L1 penalty for feature selection
- Can shrink coefficients to exactly zero (automatic feature selection)
- Creates sparse models
- Can be unstable when features are highly correlated
- Tends to select one feature from a group of correlated features

In [20]:
# Lasso
reg = Lasso(alpha=0.01)
reg.fit(X_train,y_train)
y_pred = reg.predict(X_test)
r2_score(y_test,y_pred)

0.44111855963110613

### 4. ElasticNet Regression - Complete Mathematical Derivation

## ElasticNet Cost Function

ElasticNet combines both L1 and L2 regularization penalties. We'll show this in **two formats**:

### **Sigma (Summation) Format:**
$$J(\theta) = \frac{1}{2m} \sum_{i=1}^{m} (h_\theta(x^{(i)}) - y^{(i)})^2 + \alpha \rho \sum_{j=1}^{n} |\theta_j| + \frac{\alpha(1-\rho)}{2} \sum_{j=1}^{n} \theta_j^2$$

### **Matrix Format:**
$$J(\theta) = \frac{1}{2m} ||X\theta - y||^2_2 + \alpha \rho ||\theta||_1 + \frac{\alpha(1-\rho)}{2} ||\theta||^2_2$$

Where:
- $m$ = number of training samples (rows)
- $n$ = number of features (columns)
- $X$ = feature matrix of shape $(m \times n)$
- $y$ = target vector of shape $(m \times 1)$
- $\theta$ = weight/coefficient vector of shape $(n \times 1)$
- $\alpha$ = regularization strength (alpha parameter in sklearn)
- $\rho$ = l1_ratio in sklearn (balance between L1 and L2)
- When $\rho = 1$: Pure Lasso
- When $\rho = 0$: Pure Ridge
- When $0 < \rho < 1$: ElasticNet combination

---

## Breaking Down the Cost Function Components

### Component 1: Mean Squared Error (MSE) - Loss Term

**Sigma Format:**
$$\text{MSE} = \frac{1}{2m} \sum_{i=1}^{m} (h_\theta(x^{(i)}) - y^{(i)})^2 = \frac{1}{2m} \sum_{i=1}^{m} \left(\sum_{j=0}^{n} \theta_j x_j^{(i)} - y^{(i)}\right)^2$$

**Matrix Format:**
$$\text{MSE} = \frac{1}{2m} ||X\theta - y||^2_2 = \frac{1}{2m} (X\theta - y)^T(X\theta - y)$$

This measures how well the model fits the training data.

### Component 2: L1 Penalty (Lasso)

**Sigma Format:**
$$\text{L1 Penalty} = \alpha \rho \sum_{j=1}^{n} |\theta_j|$$

**Matrix Format:**
$$\text{L1 Penalty} = \alpha \rho ||\theta||_1$$

- Encourages sparsity (some coefficients become zero)
- Performs automatic feature selection
- Creates interpretable models

### Component 3: L2 Penalty (Ridge)

**Sigma Format:**
$$\text{L2 Penalty} = \frac{\alpha(1-\rho)}{2} \sum_{j=1}^{n} \theta_j^2$$

**Matrix Format:**
$$\text{L2 Penalty} = \frac{\alpha(1-\rho)}{2} ||\theta||^2_2 = \frac{\alpha(1-\rho)}{2} \theta^T \theta$$

- Prevents overfitting
- Handles multicollinearity
- Keeps all features but shrinks coefficients

---

## Complete Derivation of ElasticNet Optimal Weights

We'll derive the optimal weights in **both Matrix Format and Sigma Format** in parallel.

---

### **Step 1: Set Up the Optimization Problem**

**Goal:** Minimize the cost function $J(\theta)$

**Sigma Format:**
$$\min_{\theta} J(\theta) = \min_{\theta} \left[\frac{1}{2m} \sum_{i=1}^{m} \left(\sum_{j=0}^{n} \theta_j x_j^{(i)} - y^{(i)}\right)^2 + \alpha \rho \sum_{j=1}^{n} |\theta_j| + \frac{\alpha(1-\rho)}{2} \sum_{j=1}^{n} \theta_j^2\right]$$

**Matrix Format:**
$$\min_{\theta} J(\theta) = \min_{\theta} \left[\frac{1}{2m} ||X\theta - y||^2_2 + \alpha \rho ||\theta||_1 + \frac{\alpha(1-\rho)}{2} ||\theta||^2_2\right]$$

---

### **Step 2: Expand the Squared Error Term**

**Matrix Format:**
$$||X\theta - y||^2_2 = (X\theta - y)^T(X\theta - y)$$
$$= (X\theta)^T(X\theta) - (X\theta)^T y - y^T(X\theta) + y^T y$$
$$= \theta^T X^T X \theta - 2\theta^T X^T y + y^T y$$

**Sigma Format:**
$$\sum_{i=1}^{m} (h_\theta(x^{(i)}) - y^{(i)})^2 = \sum_{i=1}^{m} \left[\left(\sum_{j=0}^{n} \theta_j x_j^{(i)}\right)^2 - 2y^{(i)}\sum_{j=0}^{n} \theta_j x_j^{(i)} + (y^{(i)})^2\right]$$

**Full Cost Function (Matrix):**
$$J(\theta) = \frac{1}{2m}[\theta^T X^T X \theta - 2\theta^T X^T y + y^T y] + \alpha \rho ||\theta||_1 + \frac{\alpha(1-\rho)}{2} \theta^T \theta$$

**Full Cost Function (Sigma):**
$$J(\theta) = \frac{1}{2m}\sum_{i=1}^{m} \left[\left(\sum_{j=0}^{n} \theta_j x_j^{(i)}\right)^2 - 2y^{(i)}\sum_{j=0}^{n} \theta_j x_j^{(i)} + (y^{(i)})^2\right] + \alpha \rho \sum_{j=1}^{n} |\theta_j| + \frac{\alpha(1-\rho)}{2} \sum_{j=1}^{n} \theta_j^2$$

---

### **Step 3: Take Partial Derivative with Respect to θ**

The L1 term $|\theta_j|$ is not differentiable at 0, so we use the **subgradient**:

**Matrix Format (for differentiable parts):**
$$\frac{\partial J}{\partial \theta} = \frac{1}{m}[X^T X \theta - X^T y] + \alpha(1-\rho)\theta + \alpha \rho \cdot \text{sign}(\theta)$$

**Sigma Format (for each coefficient $\theta_j$):**
$$\frac{\partial J}{\partial \theta_j} = \frac{1}{m}\sum_{i=1}^{m} x_j^{(i)} \left(\sum_{k=0}^{n} \theta_k x_k^{(i)} - y^{(i)}\right) + \alpha(1-\rho)\theta_j + \alpha \rho \cdot \text{sign}(\theta_j)$$

Simplified:
$$\frac{\partial J}{\partial \theta_j} = \frac{1}{m}\sum_{i=1}^{m} x_j^{(i)} (h_\theta(x^{(i)}) - y^{(i)}) + \alpha(1-\rho)\theta_j + \alpha \rho \cdot \text{sign}(\theta_j)$$

Where the **sign function** is:
$$\text{sign}(\theta_j) = \begin{cases} 
+1 & \text{if } \theta_j > 0 \\ 
0 & \text{if } \theta_j = 0 \\
-1 & \text{if } \theta_j < 0 
\end{cases}$$

---

### **Step 4: Set Gradient to Zero (for non-zero coefficients)**

For $\theta_j \neq 0$:

**Matrix Format:**
$$\frac{1}{m}X^T X \theta - \frac{1}{m}X^T y + \alpha(1-\rho)\theta + \alpha \rho \cdot \text{sign}(\theta) = 0$$

**Sigma Format:**
$$\frac{1}{m}\sum_{i=1}^{m} x_j^{(i)} \left(\sum_{k=0}^{n} \theta_k x_k^{(i)} - y^{(i)}\right) + \alpha(1-\rho)\theta_j + \alpha \rho \cdot \text{sign}(\theta_j) = 0$$

---

### **Step 5: Solve for θ**

**Matrix Format:**

Rearranging:
$$X^T X \theta + m\alpha(1-\rho)\theta = X^T y - m\alpha \rho \cdot \text{sign}(\theta)$$

Factoring out $\theta$:
$$(X^T X + m\alpha(1-\rho)I)\theta = X^T y - m\alpha \rho \cdot \text{sign}(\theta)$$

Where $I$ is the $(n \times n)$ identity matrix.

**Solving for $\theta$:**

**For positive coefficients** $(\theta_j > 0, \text{sign}(\theta_j) = +1)$:
$$\theta = (X^T X + m\alpha(1-\rho)I)^{-1}(X^T y - m\alpha \rho \mathbf{1})$$

**For negative coefficients** $(\theta_j < 0, \text{sign}(\theta_j) = -1)$:
$$\theta = (X^T X + m\alpha(1-\rho)I)^{-1}(X^T y + m\alpha \rho \mathbf{1})$$

Where $\mathbf{1}$ is a vector of ones.

---

**Sigma Format:**

Rearranging for a single coefficient $\theta_j$:
$$\frac{1}{m}\sum_{i=1}^{m} x_j^{(i)} \sum_{k=0}^{n} \theta_k x_k^{(i)} - \frac{1}{m}\sum_{i=1}^{m} x_j^{(i)} y^{(i)} + \alpha(1-\rho)\theta_j + \alpha \rho \cdot \text{sign}(\theta_j) = 0$$

Separating $\theta_j$ term:
$$\frac{1}{m}\sum_{i=1}^{m} x_j^{(i)} \theta_j x_j^{(i)} + \frac{1}{m}\sum_{i=1}^{m} x_j^{(i)} \sum_{k \neq j} \theta_k x_k^{(i)} - \frac{1}{m}\sum_{i=1}^{m} x_j^{(i)} y^{(i)} + \alpha(1-\rho)\theta_j + \alpha \rho \cdot \text{sign}(\theta_j) = 0$$

$$\theta_j \left[\frac{1}{m}\sum_{i=1}^{m} (x_j^{(i)})^2 + \alpha(1-\rho)\right] = \frac{1}{m}\sum_{i=1}^{m} x_j^{(i)} \left[y^{(i)} - \sum_{k \neq j} \theta_k x_k^{(i)}\right] - \alpha \rho \cdot \text{sign}(\theta_j)$$

**Solving for $\theta_j$:**
$$\theta_j = \frac{\frac{1}{m}\sum_{i=1}^{m} x_j^{(i)} \cdot r_j^{(i)} - \alpha \rho \cdot \text{sign}(\theta_j)}{\frac{1}{m}\sum_{i=1}^{m} (x_j^{(i)})^2 + \alpha(1-\rho)}$$

Where $r_j^{(i)} = y^{(i)} - \sum_{k \neq j} \theta_k x_k^{(i)}$ is the **residual without feature $j$**.

---

## Coordinate Descent Algorithm (Used by Sklearn)

Since the L1 penalty is not differentiable at zero, ElasticNet is typically solved using **Coordinate Descent**. We optimize one coefficient at a time while holding others fixed.

### Algorithm:

1. **Initialize** $\theta = \theta^{(0)}$ (e.g., all zeros or random values)

2. **Repeat** until convergence:
   
   For each feature $j = 1, 2, ..., n$:
   
   **a. Compute the residual without feature $j$:**
   
   **Sigma Format:**
   $$r_j^{(i)} = y^{(i)} - \sum_{k \neq j} \theta_k x_k^{(i)} \quad \text{for } i = 1, 2, ..., m$$
   
   **Matrix Format:**
   $$r_j = y - \sum_{k \neq j} X_k \theta_k$$
   
   **b. Compute the correlation/gradient:**
   
   **Sigma Format:**
   $$\rho_j = \sum_{i=1}^{m} x_j^{(i)} \cdot r_j^{(i)}$$
   
   **Matrix Format:**
   $$\rho_j = X_j^T r_j$$
   
   **c. Apply soft-thresholding (L1) and shrinkage (L2):**
   
   **Sigma Format:**
   $$\theta_j = \frac{\text{SoftThreshold}(\rho_j, m\alpha\rho)}{\sum_{i=1}^{m} (x_j^{(i)})^2 + m\alpha(1-\rho)}$$
   
   **Matrix Format:**
   $$\theta_j = \frac{\text{SoftThreshold}(\rho_j, m\alpha\rho)}{X_j^T X_j + m\alpha(1-\rho)}$$
   
   Where the **Soft-Thresholding operator** is:
   $$\text{SoftThreshold}(z, \lambda) = \begin{cases} 
   z - \lambda & \text{if } z > \lambda \\
   0 & \text{if } |z| \leq \lambda \\
   z + \lambda & \text{if } z < -\lambda
   \end{cases}$$
   
   This operator performs:
   - **Feature selection** (L1): Sets coefficient to 0 if correlation is small
   - **Shrinkage** (L2): Reduces coefficient magnitude

3. **Check convergence**: If $||\theta^{(t+1)} - \theta^{(t)}||_2 < \epsilon$, stop.

---

## Final Weight Update Formula

For each coefficient $\theta_j$, we have two equivalent formulations:

### **Matrix Format:**
$$\theta_j = \frac{\text{SoftThreshold}(X_j^T r_j, m\alpha\rho)}{||X_j||^2_2 + m\alpha(1-\rho)}$$

### **Sigma Format:**
$$\theta_j = \frac{\text{SoftThreshold}\left(\sum_{i=1}^{m} x_j^{(i)} \cdot r_j^{(i)}, m\alpha\rho\right)}{\sum_{i=1}^{m} (x_j^{(i)})^2 + m\alpha(1-\rho)}$$

Where:
- **Residual without feature $j$**: 
  - Sigma: $r_j^{(i)} = y^{(i)} - \sum_{k \neq j} \theta_k x_k^{(i)}$
  - Matrix: $r_j = y - X_{-j}\theta_{-j}$
- **Feature norm squared**: 
  - Sigma: $\sum_{i=1}^{m} (x_j^{(i)})^2$
  - Matrix: $||X_j||^2_2 = X_j^T X_j$

### **Detailed Step-by-Step Update (Sigma Format):**

1. **Compute residual for all samples:**
   $$r_j^{(i)} = y^{(i)} - \sum_{k \neq j} \theta_k x_k^{(i)}, \quad i = 1, ..., m$$

2. **Compute correlation:**
   $$z = \sum_{i=1}^{m} x_j^{(i)} \cdot r_j^{(i)}$$

3. **Apply soft-thresholding:**
   $$z_{\text{soft}} = \begin{cases} 
   z - m\alpha\rho & \text{if } z > m\alpha\rho \\
   0 & \text{if } |z| \leq m\alpha\rho \\
   z + m\alpha\rho & \text{if } z < -m\alpha\rho
   \end{cases}$$

4. **Normalize with L2 penalty:**
   $$\theta_j = \frac{z_{\text{soft}}}{\sum_{i=1}^{m} (x_j^{(i)})^2 + m\alpha(1-\rho)}$$

---

## Key Parameters in Scikit-learn

```python
ElasticNet(alpha=1.0, l1_ratio=0.5)
```

- **alpha** ($\alpha$): Regularization strength
  - Larger values = more regularization = simpler models
  - Default: 1.0

- **l1_ratio** ($\rho$): Mix ratio between L1 and L2
  - l1_ratio = 0: Pure Ridge (L2)
  - l1_ratio = 1: Pure Lasso (L1)
  - l1_ratio = 0.5: Equal mix
  - Default: 0.5

---

## When to Use ElasticNet?

✅ **Use ElasticNet when:**
- You have many features with high correlation
- You want automatic feature selection (like Lasso)
- You want stability with correlated features (like Ridge)
- You're dealing with high-dimensional data (n > m)
- You want sparse models with grouped variable selection

❌ **Avoid ElasticNet when:**
- You have few features with no correlation (use Ridge or Lasso)
- Computational cost is a major concern (ElasticNet is slower)
- You need the fastest possible training (use Ridge)

In [21]:
# ElasticNet
reg = ElasticNet(alpha=0.005,l1_ratio=0.9)
reg.fit(X_train,y_train)
y_pred = reg.predict(X_test)
r2_score(y_test,y_pred)

0.4531474541554823

---

## Model Comparison Summary

Let's compare the R² scores of all four models:

| Model | Regularization | R² Score | Key Characteristics |
|-------|----------------|----------|---------------------|
| **Linear Regression** | None | ~0.52 | Baseline, no regularization |
| **Ridge** | L2 | ~0.45 | Shrinks coefficients, keeps all features |
| **Lasso** | L1 | ~0.46 | Feature selection, can set coefficients to zero |
| **ElasticNet** | L1 + L2 | ~0.46 | Combines benefits of both |

### Key Observations:

1. **Linear Regression**: Highest R² on training data but may overfit
2. **Ridge**: Good for handling multicollinearity, stable predictions
3. **Lasso**: Performs feature selection, creates sparse models
4. **ElasticNet**: Most flexible, handles correlated features better than Lasso

### Hyperparameter Tuning Tips:

For ElasticNet, you need to tune two parameters:
- **alpha**: Start with values like [0.001, 0.01, 0.1, 1, 10, 100]
- **l1_ratio**: Try values like [0.1, 0.3, 0.5, 0.7, 0.9, 1.0]

```python
# Example hyperparameter search
from sklearn.model_selection import GridSearchCV

param_grid = {
    'alpha': [0.001, 0.01, 0.1, 1, 10],
    'l1_ratio': [0.1, 0.3, 0.5, 0.7, 0.9]
}

elastic_net = ElasticNet()
grid_search = GridSearchCV(elastic_net, param_grid, cv=5, 
                          scoring='r2', n_jobs=-1)
grid_search.fit(X_train, y_train)
```

---

## Conclusion

**ElasticNet** is a powerful regularization technique that:
- ✅ Combines L1 and L2 penalties for better regularization
- ✅ Performs feature selection like Lasso
- ✅ Handles correlated features better than Lasso
- ✅ More stable than Lasso in presence of multicollinearity
- ✅ Flexible via the l1_ratio parameter

**Use ElasticNet when** you have many correlated features and want both feature selection and model stability!